fmt/mod.c

```diff
@@ -396,3 +467,192 @@ int fmt_mod15_load_song(song_t *song, slurp_t *fp, unsigned int lflags)
 {
 	return fmt_mod_load_song(song, fp, lflags, 1);
 }
+
+/* incomplete 31 sample M.K. amiga mod saving routine */
+int fmt_mod_save_song(disko_t *fp, song_t *song)
+{
+	uint8_t mod_songtitle[20];
+	uint8_t mod_sampleheader[30];
+	uint8_t mod_orders[128];
+	uint8_t tmp[128];
+	uint8_t mod_pattern[1024];
+
+	int nord, nsmp, maxpat, jmax, joutpos;
+	int i, j, n, period;
+	unsigned int warn = 0;
+	song_note_t *m;
+
+	if (song->flags & SONG_INSTRUMENTMODE)
+		warn |= 1 << WARN_INSTRUMENTS;
+	if (song->flags & SONG_LINEARSLIDES)
+		warn |= 1 << WARN_LINEARSLIDES;
+
+	nsmp = csf_get_num_samples(song); // Getting number of samples
+	if (nsmp > 31) {
+		nsmp = 31;
+		warn |= 1 << WARN_MAXSAMPLES;
+	}
+
+	if (3 < csf_get_highest_used_channel(song))
+		warn |= 1 << WARN_MAXCHANNELS;
+
+	memcpy(mod_songtitle, song->title, 20);
+	disko_write(fp, mod_songtitle, 20); // writing song title
+
+	// Now writing sample headers
+	for(n = 1; n <= 31; ++n) {
+		for(i = 0; i < 30; ++i) mod_sampleheader[i] = 0;
+		if(n <= nsmp) {
+			if(song->samples[n].global_volume != 64)
+				warn |= 1 << WARN_SAMPLEVOL;
+			if((song->samples[n].flags & (CHN_LOOP | CHN_PINGPONGLOOP)) == (CHN_LOOP | CHN_PINGPONGLOOP) || (song->samples[n].flags & CHN_SUSTAINLOOP))
+				warn |= 1 << WARN_LOOPS;
+			if(song->samples[n].vib_depth != 0)
+				warn |= 1 << WARN_SAMPLEVIB;
+			memcpy(mod_sampleheader, song->samples[n].name, 22); // sample name
+			mod_sampleheader[22] = song->samples[n].length >> 9; // sample 11th word MSB length/2
```

Comment:
This should limit the sample length to 0x1FFFE samples before shifting. Otherwise longer samples will be truncated. Loop start / length should be adjusted accordingly if they are beyond the sample length.

Reply:
csf_write_sample() updated with truncate-and-mask argument, which is ignored when it is uint32_max